frontend/src/utils.js

```diff
@@ -0,0 +1,122 @@
+export const timeAgo = (date, displayPrefix = false) => {
+  if (typeof date === "string") {
+    date = new Date(Date.parse(date))
+  }
+
+  let prefix = ""
+  if (displayPrefix) {
+    prefix = "il y a "
+  }
+
+  const seconds = (new Date() - date) / 1000
+  if (seconds < 60) {
+    return "à l'instant"
+  }
+  const minutes = Math.round(seconds / 60)
+  if (minutes < 60) {
+    return `${prefix}${minutes} m`
+  }
+  const hours = Math.round(seconds / 3600)
+  if (hours < 24) {
+    return `${prefix}${hours} h`
+  }
+  const days = Math.round(seconds / 3600 / 24)
+  if (days < 7) {
+    return `${prefix}${days} j`
+  }
+
+  prefix = "le "
+
+  if (days < 120) {
+    return prefix + date.toLocaleString("fr-FR", { month: "long", day: "numeric" })
+  }
+  return prefix + date.toLocaleString("fr-FR", { month: "long", day: "numeric", year: "numeric" })
+}
+export const toBase64 = (file, success, error) => {
+  const reader = new FileReader()
+  reader.readAsDataURL(file)
+  reader.onload = function() {
+    success(reader.result)
+  }
+  if (error) reader.onerror = error
+}
+
+function arraysMatch(arr1, arr2) {
+  if (arr1.length !== arr2.length) return false
+
+  for (var i = 0; i < arr1.length; i++) {
+    if (arr1[i] !== arr2[i]) return false
+  }
+
+  return true
+}
+
+export const getObjectDiff = (obj1, obj2) => {
+  if (!obj2 || Object.prototype.toString.call(obj2) !== "[object Object]") {
+    return obj1
+  }
+
+  let diffs = {}
+  let key
+
+  let compare = function(item1, item2, key) {
+    var type1 = Object.prototype.toString.call(item1)
+    var type2 = Object.prototype.toString.call(item2)
+
+    // If type2 is undefined it has been removed
+    if (type2 === "[object Undefined]") {
+      diffs[key] = null
+      return
+    }
+
+    // If items are different types
+    if (type1 !== type2) {
+      diffs[key] = item2
+      return
+    }
+
+    // If an object, compare recursively
+    if (type1 === "[object Object]") {
+      var objDiff = getObjectDiff(item1, item2)
+      if (Object.keys(objDiff).length > 0) {
+        diffs[key] = objDiff
+      }
+      return
+    }
+
+    // If an array, compare
+    if (type1 === "[object Array]") {
+      if (!arraysMatch(item1, item2)) {
+        diffs[key] = item2
+      }
+      return
+    }
+
+    // If it's a function, convert to a string and compare
+    if (type1 === "[object Function]") {
+      if (item1.toString() !== item2.toString()) {
+        diffs[key] = item2
+      }
+    } else {
+      if (item1 !== item2) {
+        diffs[key] = item2
+      }
+    }
+  }
+
+  for (key in obj1) {
+    if (Object.prototype.hasOwnProperty.call(obj1, key)) {
+      compare(obj1[key], obj2[key], key)
+    }
+  }
+
+  for (key in obj2) {
+    if (Object.prototype.hasOwnProperty.call(obj2, key)) {
+      if (!obj1[key] && obj1[key] !== obj2[key]) {
```

Comment:
Pourquoi est-ce qu'on a besoin de `!obj1[key]` ?

Reply:
Car si `key` existe dans le premier objet, cette clé aura déjà été traité dans la boucle précédente, celle qui itère le premier objet :
 
```javascript
  // Ici on traitera toutes les clés du premier objet
  for (key in obj1) {
    if (Object.prototype.hasOwnProperty.call(obj1, key)) {
      compare(obj1[key], obj2[key], key)
    }
  }

  for (key in obj2) {
    if (Object.prototype.hasOwnProperty.call(obj2, key)) {
      // Si la clé est présente dans le premier objet, pas besoin de comprarer car on l'a déjà fait 
      if (!obj1[key] && obj1[key] !== obj2[key]) {
        diffs[key] = obj2[key]
      }
    }
  }
```